### PR TITLE
Implicit GEMM 3D backward data V4R1 and V1R1

### PR DESCRIPTION
**Pull request by asleepzzz**
Implicit GEMM 3D backward data V4R1 and V1R1

### CODE REVIEW

**Comment by asleepzzz** [comment]
[gfx906_60.HIP.2_4_0.ufdb.txt](https://github.com/ROCmSoftwarePlatform/MIOpen/files/4468474/gfx906_60.HIP.2_4_0.ufdb.txt)

**Comment by asleepzzz** [comment]
[3d_conv_fwd_bwd.xlsx](https://github.com/ROCmSoftwarePlatform/MIOpen/files/4468484/3d_conv_fwd_bwd.xlsx)
table "resnet3d bwd vs gemm"

**Comment by daniellowell** [comment]
All, reviews please.

File: src/ocl/convolutionocl.cpp [hunk 1]
```diff
@@ -245,12 +248,61 @@ EvaluateDataImplicitGemmSolution(Handle& handle,
...
+        { // only pad=0 stride=1 filter 1 not set zero
+            bool is_1x1_s1 = false;
+            if(miopen::all_of(padding, [](auto v) { return v == 0; }) &&
+               miopen::all_of(strides, [](auto v) { return v == 1; }))
+            {
+                if(wDesc.GetLengths()[2] == 1 && wDesc.GetLengths()[3] == 1)
+                { // filter = 1
+                    if(wDesc.GetSize() == 4 || (wDesc.GetSize() == 5 && wDesc.GetLengths()[4] == 1))
+                    {
+                        is_1x1_s1 = true;
```
Comment (asroy) [inline src/ocl/convolutionocl.cpp:263]: [NOT URGENT] ```is_1x1_s1``` the name is not correct. You are not checking if it's 1x1 and s1, you are checking if the stride is larger than filter size. Please fix the naming in the future PR

File: src/ocl/convolutionocl.cpp [hunk 2]
```diff
@@ -245,12 +248,61 @@ EvaluateDataImplicitGemmSolution(Handle& handle,
...
+        else if(kernel.GetName() ==
+                    "gridwise_convolution_backward_data_implicit_gemm_v4r1_nchw_kcyx_nkhw" ||
+                kernel.GetName() ==
+                    "gridwise_convolution_backward_data_implicit_gemm_v4r1_ncdhw_kczyx_nkdhw")
         {
-            // \todo this kernel doesn't always need to set-zero
-            float zero = 0.f;
-            SetTensor(handle, outDesc, out, &zero);
-            elapsed += handle.GetKernelTime();
+            bool is_1x1_s1 = false;
```
Comment (asroy) [inline src/ocl/convolutionocl.cpp:280]: [NOT URGENT] same here, please fix the naming in the future

File: src/ocl/convolutionocl.cpp [hunk 3]
```diff
@@ -3039,14 +3097,73 @@ void ConvBwdImplicitGemm(const ConvolutionContext& /*ctx*/,
...
-            "gridwise_convolution_backward_data_implicit_gemm_v4r1_nchw_kcyx_nkhw")
+                "gridwise_convolution_backward_data_implicit_gemm_v1r1_nchw_kcyx_nkhw" ||
+            kernel.GetName() ==
+                "gridwise_convolution_backward_data_implicit_gemm_v1r1_ncdhw_kczyx_nkdhw")
     {
-        // \todo this kernel doesn't always need to set-zero
-        float zero = 0.f;
-        SetTensor(handle, tensors.dxDesc, tensors.dx, &zero);
+        // this kernel accumulate results into input tensor, therefore need to set zero
+        bool is_1x1_s1 = false;
```
Comment (asroy) [inline src/ocl/convolutionocl.cpp:3105]: [NOT URGENT] is_1x1_s1 naming is wrong

File: src/ocl/convolutionocl.cpp [hunk 4]
```diff
@@ -3039,14 +3097,73 @@ void ConvBwdImplicitGemm(const ConvolutionContext& /*ctx*/,
...
+            { // filter = 1
+                if(tensors.wDesc.GetSize() == 4 ||
+                   (tensors.wDesc.GetSize() == 5 && tensors.wDesc.GetLengths()[4] == 1))
+                {
+                    is_1x1_s1 = true;
+                }
+            }
+        }
 
+        if(!is_1x1_s1)
```
Comment (asroy) [inline src/ocl/convolutionocl.cpp:3119]: [NOT URGENT] is_1x1_s1 naming is wrong

File: src/ocl/convolutionocl.cpp [hunk 5]
```diff
@@ -3039,14 +3097,73 @@ void ConvBwdImplicitGemm(const ConvolutionContext& /*ctx*/,
...
         if(handle.IsProfilingEnabled())
             elapsed += handle.GetKernelTime();
+    }
+    else if(kernel.GetName() ==
+                "gridwise_convolution_backward_data_implicit_gemm_v4r1_nchw_kcyx_nkhw" ||
+            kernel.GetName() ==
+                "gridwise_convolution_backward_data_implicit_gemm_v4r1_ncdhw_kczyx_nkdhw")
+    {
+        // \todo this kernel doesn't always need to set-zero
+        bool is_1x1_s1 = false;
```
Comment (asroy) [inline src/ocl/convolutionocl.cpp:3137]: [NOT URGENT] is_1x1_s1 naming is wrong

File: src/ocl/convolutionocl.cpp (hunk 2 above)
Comment (asleepzzz) [reply src/ocl/convolutionocl.cpp]: v4r1 condition change to filterGeStride

**Comment by atamazov** [comment]
@asroy 
> In future PR, please fix the naming for is_1x1_s1. You are checking if the filter size is larger than stride, not if it's 1x1 stride1

Is this PR urgent? None priority labels set.

Please do not allow merging PRs with problems unless absolutely necessary (blockers). We allowed that in the past, but I see that decision was wrong. Many posptponed review comments remain unresolved. It seems like the responsible programmers are doing other things, in spite of that the agreement was to resolve the leftovers as if these have `priority_high` (and the appropriate ticket is labeled so). Perhaps they are always doing the blocker PRs since then, but it's hard to believe in that.

/cc @daniellowell

**Comment by asleepzzz** [comment]
Hi atam,
I will remove some additional () you comments in last PR.
Due to this PR is important and have a lot of tasks so I implement this at first priority.
Sorry for late to reply your comments

**Comment by asleepzzz** [comment]
I will create another PR to remove additional () in 3d fwd

**Comment by atamazov** [comment]
@asleepzzz 

> Hi atam,
> I will remove some additional () you comments in last PR.
> Due to this PR is important and have a lot of tasks so I implement this at first priority.
> Sorry for late to reply your comments

No problem. That comment is addressed to Daniel and Chao.
--Artem

**Comment by asleepzzz** [comment]
db already provided 10 days ago, do you need more configs?

**Pull request merged**